src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py

```diff
@@ -0,0 +1,599 @@
+        self._sparsity = sparsity
+        self._sparsity_strategy = sparsity_strategy
+        self._update_fraction = None
+        # compute sparsity distribution
+        self._sparsity_distribution = self.get_sparsity_distribution()
+        # prepare grad storage
+        self._param_grads = [None] * len(params)
+        self._enabled_grad_buffering = False
+
+    def sparsity_scaler(self, score: Tensor) -> float:
+        """
+        Assigns the sparsity scale for a given parameter
+        according to the sparsity_strategy. The weights
+        with smaller scale are pruned more than those with larger.
+        """
+        assert len(score.shape) >= 2, "Pruned weight must be at least 2-dimensional."
+        if self._sparsity_strategy == "uniform":
+            return 1.0
+        elif self._sparsity_strategy == "erdos_renyi":
+            c_out, c_in = score.shape[:2]
+            return (c_in + c_out) / (c_in * c_out)
+        elif self._sparsity_strategy == "erdos_renyi_kernel":
+            return np.sum(score.shape) / np.prod(score.shape)
+        else:
+            raise ValueError("Unknown sparsity distribution")
+
+    @torch.no_grad()
+    def get_sparsity_distribution(self) -> None:
+        """
+        Returns the sparsity distribution for every parameter
+        scaled with respect to the sparsity strategy.
+        """
+        total_params = 0
+        cumulative_sum = 0
+        for param in self._params:
+            cumulative_sum += self.sparsity_scaler(param) * param.numel()
+            total_params += param.numel()
+        norm_factor = ((1 - self._sparsity) * total_params) / cumulative_sum
+        return [
+            np.clip(1 - norm_factor * self.sparsity_scaler(param), 0.0, 1.0)
+            for param in self._params
+        ]
+
+    @torch.no_grad()
+    def pre_optim_step_update(self, masks: List[Tensor]):
+        """
+        Collect single gradient for estimation of RigL saliency score.
+
+        :param masks: latest masks that are applied to these parameters
+        """
+        if not self._enabled_grad_buffering:
+            # only collect gradients when called during pruning step
+            # this ignores calls invoked by manager during training
+            return
+
+        for i, param_grad in enumerate(self._param_grads):
+            if param_grad is None:
+                self._param_grads[i] = self._params[i].grad.clone()
+            else:
+                self._param_grads[i].add_(self._params[i].grad)
+
+    def get_param_score(self, param: Tensor, param_grad: Tensor, param_sparsity: float):
+        """
+        Computed the saliency score for a given parameter.
+
+        :param param: param to be scored
+        :param param_grad: gradient of the parameters
+        :param param_sparsity: the sparsity for a given parameter
+        """
+        # We drop and replace update_fraction of nonzero elements
+        updated_number = (1 - param_sparsity) * self._update_fraction
```

Comment:
It's not a decrement, because sparsity is always constant. Rather, it's the proportion of the mask that can also change during the update. I changed it to mask_update_fraction.